### DISCUSSION TITLE
Is Maui ready for Production?

### DISCUSSION THREAD

**Discussion by sikemullivan** [dotnet/maui#22844] [General] [answered]
I've been building an application in Maui for about 5 months. Prior to this app, I've built over 10 mobile apps mostly using React Native.  I was really impressed after reading and watching a lot of material on Maui. Since I'm also a dotnet developer, it seemed like a no brainer to use Maui. 

I'm just curious if anyone has built a production grade application with Maui? I've been able to build my app, and make it do what I want it to do. However, it is loaded with bugs and I'd never present this to anybody (not even my wife). The CollectionView is a crash generator. Sad thing is, I haven't even tested my app with Android or iOS yet. I'm struggling to solve bugs even in Windows. 

I'm not sure what to do. Am I using Maui incorrectly? I feel like Maui isn't ready for production. Does an actual production app exist out there? Can I use it as an example?  

At this point, I feel like this repo should warn developers that Maui isn't ready. If you're looking to build a production application, use something else. Maui is for pretending you're building a Mobile app in C#.

**Comment by MartyIX** [comment, marked as answer]
I can probably answer a few questions but not all.

> I'm not sure what to do. Am I using Maui incorrectly?

You most likely use it correctly. 

> I feel like Maui isn't ready for production.

Yes, in my opinion, there are still _much more_ bugs than expected.

> Does an actual production app exist out there? Can I use it as an example?

There are some samples here: https://github.com/jsuarezruiz/awesome-dotnet-maui/?tab=readme-ov-file#samples But I suppose this is not what you are looking for but I don't know about a better link.

> At this point, I feel like this repo should warn developers that Maui isn't ready.

It gets better with every release. IMO the pace of bug fixing gets better because the testing architecture is in much better shape than a few months ago. But yes, if you need a rock solid platform MAUI is not there (yet). One can watch MAUI SDK Ongoing board (https://github.com/orgs/dotnet/projects/194/views/31) to see if his/her bug is in the pipeline to get fixed.

> If you're looking to build a production application, use something else.

There are many questions and personal preferences in this decision. Like how complex your app is. Do you want to spend your time working with React or do you want to use only .NET, etc. etc. But yes, a warning is a fair thing to do.

> I'm struggling to solve bugs even in Windows.

Support for Windows was added the last. In my opinion, it gets less attention than Android nad iOS because the first priority is to allow people migrate from Xamarin to MAUI. But it's [certainly possible](https://github.com/dotnet/maui/pulls?q=is%3Apr+author%3AMartyIX+is%3Aclosed) to fix Windows issues in MAUI.

**Reply by gautambjain** [reply]
> Support for Windows was added the last. In my opinion, it gets less attention than Android and iOS because the first priority is to allow people migrate from Xamarin to MAUI.

Yes I agree with this. I can tell that Android and iOS apps are quite production ready. Not sure of Windows apps.

**Reply by sikemullivan** [reply]
Thanks @MartyIX and @gautambjain, I didn't realize Windows support was last. I was using Windows as my baseline for development because it was "easier". I guess that was a mistake on my part . Going forward, I'll stick with using the Android emulator.

**Comment by kj4bix** [comment]
I have an application in the stores developed using .NET MAUI. The app uses most everything you can on a phone. Location services, mapping, azure mobile datasync, local SQLite databases, 3d graphic display, bar code scanning, being able to work entirely disconnected and sync when you have a connection. Frankly, I thought it was rather buggy with .NET 6.0, but a BUNCH of fixes were made between 6.0 and 8.0. Yes, there are items that still don't work as you would expect, but in general it works quite well. My advice would be don't hang on to the old Xamarin.Forms schemes of a bunch of nested StackLayouts, in general move to Grids - yes, I even have Grids with a single row due to layout issues with other approaches, use the community toolkit, use their MVVM. Don't use items like a ListView use CollectionViews instead, don't use Frames, use Borders, there are a bunch of gotchas like that, but it does work. I started Xamarin.Forms when it was beta - frankly it was far buggier than 8.0 is today. I'm writing a second app now and it is going far quicker than the first one which took around a year to replace the original Xamarin.Forms application.  It runs on iOS, Android and Windows.

**Comment by kiddailey** [comment]
TLDR - For a simple app, probably.  For anything beyond that, I'm not so sure.

Full disclosure -- I come from years of Xamarin development and hold it in high regards.  I've been excited about MAUI and eager to migrate for a variety of reasons (simplified project structure, speed improvements, modernization, etc).  I have a production Xamarin app that I've been trying to migrate to MAUI.  I also have a separate project that was started in MAUI.

The Xamarin app is feature rich, uses a variety of technologies and is *very* stable on iOS and Android.  Xamarin (even with its warts) feels mature and solid.    Porting the app to MAUI has been slow-going and challenging due to issues with the platform and/or 3rd-party support.

My new MAUI project started easily, but I discovered that when I wanted to deviate from simple things, the lack of polish becomes evident.  I have spent hours combing the issues list, documentation and the internet trying to figure out fixes and workarounds.  So much so that I've been eyeing other potential platforms/frameworks.

The biggest concern I have (and this also existed during Xamarin development for what it's worth) is that bugs get reported, only to be automatically closed without ever being addressed or being added to the growing "Backlog Milestone."   The automatic bot has a bad habit of closing issues due to inactivity and not because they were resolved.

I'm sure it'll eventually get to a point where I can say "YES" with confidence, but I think that's going to take a while yet.
